cmd/diff.go

```diff
@@ -102,3 +118,183 @@ files.`,
   # referred to by './kubeconfig'.
   ks diff --kubeconfig=./kubeconfig -f ./pod.yaml`,
 }
+
+func initDiffCmd(cmd *cobra.Command, wd metadata.AbsPath, envFq1, envFq2 *string, files []string, diffStrategy string) (kubecfg.DiffCmd, error) {
+	const (
+		remote = "remote"
+		local  = "local"
+	)
+
+	var err error
+
+	// ---------------------------------------------------------------------------
+	// Diff between expanded Kubernetes objects and objects on a remote cluster
+	// ---------------------------------------------------------------------------
+	if envFq2 == nil {
+		c := kubecfg.DiffRemoteCmd{}
+		c.DiffStrategy = diffStrategy
+		c.Client = &kubecfg.Client{}
+
+		if envFq1 != nil && (strings.HasPrefix(*envFq1, remote+":") || strings.HasPrefix(*envFq1, local+":")) {
+			return nil, fmt.Errorf("single <env> argument with prefix '%s': or '%s': not allowed", local, remote)
+		}
+
+		envSpec := &envSpec{env: envFq1, files: files}
+
+		c.Client.APIObjects, err = expandEnvCmdObjs(cmd, envSpec, wd)
```

Comment:
Ok, so I think the pattern here is that these three types of diff involve some combination of the following two cases:

1. Expand local version of the app into Kubernetes API objects, pass those to `kubecfg.*`
1. Get a client pool, pass to `kubecfg.*` and have that retrieve the API objects from the server.

In other commands, the API objects are always expanded before passing to `kubecfg.*`, and the reason is to make the interfaces testable. (You may remember Gus complained and we had to basically revert.) This will become more important as we have more unit tests to substantiate diffs based on the persistent cache of previous app states.

So: I suspect but do not know that it would be simpler to have a single `kubecfg.DiffCmd` type that contains two sets of API objects and a diff strategy, and a function `Run` that simply executes the diff, which would be much more similar in principle to how (_e.g._) `apply` works.

Assuming this makes sense, I would also break this monofunction into smaller sub-functions (as mentioned above).

Reply:
I'll break up `initDiffCmd()` into smaller functions but i'm not sure that having a single `kubecfg.DiffCmd` taking two sets of API objects and a diff strategy is the right option for a few reasons --

1. We'd be coupling the logic for retrieving live objects further with the `cmd/` package, which is already doing much more than adding commands and command line parsing.
2. If it's the unit testability of diffing between two sets of API objects that is a concern, testing the `diffAll` function should capture this. It's a standalone function performing the diff that takes two sets of API objects as it's main params. 
3. I worry we wouldn't be capturing the state of things -- that we are in fact supporting multiple diff command types; diffing between objects locally, between objects on server, and between objects on server and locally. I feel that the application logic layer should be capturing this in the `kubecfg` package instead of hidden in `cmd`.

Reply:
Ok, it turns out `kubecfg.DiffCmd` is currently fetching from the server, so I'm fine with this staying as it is.